### PR TITLE
feat(identity-rs):  add create identity

### PR DESCRIPTION
**Pull request by MatejVukosav**
- Initial setup of identity repo
- method to create an identity from the public key 
- types following document structure

In example I am creating keypair with the same method as in chat-p0c

### CODE REVIEW

File: identity-rs/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+libp2p = "0.53.2"
+libc = "0.2.110"
+didkit-cli = "0.3.2"
+serde_json = "1.0.113"
+sha2 = "0.10.8"
+ed25519-dalek = "2.1.1"
+multibase = "0.9.1"
+hex = "0.4.3"
+bs58 = "0.5.0"
+serde = "1.0.196"
```
Comment (miraclx) [inline identity-rs/Cargo.toml]: A bunch of the deps here aren't used, can you clean them up?

File: identity-rs/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+
+
+# Decentralized identity, also referred to as self-sovereign identity, is an open-standards based identity framework that uses digital identifiers and verifiable credentials that are self-owned, independent, and enable trusted data exchange.
```
Comment (miraclx) [inline identity-rs/README.md]:
```suggestion
# Calimero Identity

Decentralized identity also referred to as self-sovereign identity, is an open-standards-based identity framework that uses digital identifiers and verifiable credentials that are self-owned, independent, and enable trusted data exchange.
```

File: identity-rs/src/identity_provider.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+///       "publicKeyMultibase": "zCovLVG4fQcqT8sDqj76uQuXtQU2LqABAf6X8vnDW36zAidisFK22Z5Ecm28apKb9Kg6ofRo",
+///       "controller": "did:cali:12D3KooWLU42rBMLrzAyFg8CdpHiPcwJcVsKv9Wx9DtiT4QjPGGV"
+///     }
+///   ]
+/// }
+pub fn create_identity(authentication: Authentication) -> DidDocument {
+    let public_key_id = authentication.public_key.to_peer_id().to_base58();
+    let multibase_encoded = encode(Base::Base58Btc, &public_key_id);
+
+    let did = DID_CALI_IDENTIFIER.to_string() + &public_key_id;
```
Comment (miraclx) [inline identity-rs/src/identity_provider.rs]:
```suggestion
    let did = format!("{}{}", DID_CALI_IDENTIFIER, public_key_id);
```

File: identity-rs/src/identity_provider.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+///       "id": "did:cali:12D3KooWLU42rBMLrzAyFg8CdpHiPcwJcVsKv9Wx9DtiT4QjPGGV",
+///       "type": "Ed25519",
+///       "publicKeyMultibase": "zCovLVG4fQcqT8sDqj76uQuXtQU2LqABAf6X8vnDW36zAidisFK22Z5Ecm28apKb9Kg6ofRo",
+///       "controller": "did:cali:12D3KooWLU42rBMLrzAyFg8CdpHiPcwJcVsKv9Wx9DtiT4QjPGGV"
+///     }
+///   ]
+/// }
+pub fn create_identity(authentication: Authentication) -> DidDocument {
+    let public_key_id = authentication.public_key.to_peer_id().to_base58();
+    let multibase_encoded = encode(Base::Base58Btc, &public_key_id);
```
Comment (miraclx) [inline identity-rs/src/identity_provider.rs]: [`PeerId` impls `AsRef<Multihash<u64>>`](https://docs.rs/libp2p-identity/0.2.8/libp2p_identity/struct.PeerId.html#impl-AsRef%3CMultihash%3C64%3E%3E-for-PeerId) and [`Multihash<u64>` has a method `to_bytes`](https://docs.rs/multihash/0.19.1/multihash/struct.Multihash.html#method.to_bytes) that's non-consuming.

Also, [`multibase::encode`](https://docs.rs/multibase/0.9.1/multibase/fn.encode.html) uses the `base` you selected to encode the `input`, which is effectively the same as what you already did with [`.to_base58()`](https://docs.rs/libp2p-identity/0.2.8/src/libp2p_identity/peer_id.rs.html#119-121). So it becomes `base58(base58())`.

Factoring all these, we can instead write this as:

```suggestion
    let public_key_id = authentication.public_key.to_peer_id();
    let multibase_encoded = encode(Base::Base58Btc, &public_key_id.as_ref().to_bytes());
```

We would've been able to simplify to..

```diff
-    let multibase_encoded = encode(Base::Base58Btc, &public_key_id.as_ref().to_bytes());
+    let multibase_encoded = encode(Base::Base58Btc, &public_key_id.to_bytes());
```

.. but unfortunately, the [`PeerId::to_bytes`](https://docs.rs/libp2p-identity/0.2.8/libp2p_identity/struct.PeerId.html#method.to_bytes) method unnecessarily consumes the `PeerId`.

File: identity-rs/src/identity_provider.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    let did = DID_CALI_IDENTIFIER.to_string() + &public_key_id;
+
+    let verification_method: VerificationMethod = VerificationMethod {
+        id: did.clone()+"#key1",
```
Comment (miraclx) [inline identity-rs/src/identity_provider.rs]:
```suggestion
        id: format!("{}#key1", did),
```

File: identity-rs/src/identity_provider.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        algorithm_type: authentication.algorithm.to_string(),
+        public_key_multibase: multibase_encoded,
+        controller: authentication.controller.unwrap_or(did.clone()),
```
Comment (miraclx) [inline identity-rs/src/identity_provider.rs]: Allocate only when absolutely necessary

```suggestion
        controller: authentication.controller.unwrap_or_else(|| did.clone()),
```

File: identity-rs/src/identity_provider.rs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    };
+
+    let did_document: DidDocument = DidDocument {
+        id: did.clone(),
```
Comment (miraclx) [inline identity-rs/src/identity_provider.rs]: There are no further consumers of `did` so we can fully consume it here.

```suggestion
        id: did,
```

File: identity-rs/.gitignore [hunk 8]
```diff
@@ -0,0 +1,3 @@
+.idea/
+data/
+target/
```
Comment (miraclx) [inline identity-rs/.gitignore]: PS: this is duplicated from the root

File: identity-rs/src/identity_provider.rs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        verification_method: vec![verification_method],
+    };
+    //TODO store it
+    return did_document;
```
Comment (miraclx) [inline identity-rs/src/identity_provider.rs]: implicit returns :-)

```suggestion
    did_document
```

File: identity-rs/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+didkit = "0.6.0"
+libp2p = "0.53.2"
+serde_json = "1.0.113"
+sha2 = "0.10.8"
+multibase = "0.9.1"
+serde = "1.0.196"
```
Comment (miraclx) [inline identity-rs/Cargo.toml]: Can you move these definitions to the root `Cargo.toml`, you can reference them with `<package>.workspace = true`

File: identity-rs/src/types.rs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    pub algorithm_type: String,
+    #[serde(rename = "publicKeyMultibase")]
+    pub public_key_multibase: String,
+    pub controller: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct DidDocument {
+    pub id: String,
+    #[serde(rename = "verificationMethod")]
```
Comment (miraclx) [inline identity-rs/src/types.rs]: For these kinds of patterns, we can use `#[serde(rename_all = "camelCase")]` on the struct itself.

File: identity-rs/src/types.rs [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SerializedPublicKey(Vec<u8>);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct VerificationMethod {
+    pub id: String,
+    #[serde(rename = "type")]
+    pub algorithm_type: String,
+    #[serde(rename = "publicKeyMultibase")]
```
Comment (miraclx) [inline identity-rs/src/types.rs:12]: here as well

File: identity-rs/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+fn main() {
+    //generate keypair in any way
+    let keypair = Keypair::generate_ed25519();
+    let public_key = keypair.public();
+
+    let authentication: Authentication = Authentication {
+        algorithm: AlgorithmType::Ed25519,
+        controller: None,
+        public_key: public_key.clone(),
```
Comment (miraclx) [inline identity-rs/src/main.rs]:
```suggestion
        public_key,
```

File: identity-rs/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "identity-rs"
```
Comment (miraclx) [inline identity-rs/Cargo.toml]: Let's remove the `-rs` here in the crate name

File: identity-rs/src/types.rs (hunk 12 above)
Comment (MatejVukosav) [reply identity-rs/src/types.rs:12]: I cannot use it here as `algorithm_type` needs to be `type`, and `type` itself is not allowed.

Comment (miraclx) [reply identity-rs/src/types.rs:12]: That's permissible, retain the explicit rename at that point, the pattern would be:

```rust
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerificationMethod {
    pub id: String,
    #[serde(rename = "type")]
    pub algorithm_type: String,
    pub public_key_multibase: String,
```

The general idea is "default to camelCase", but use a custom-defined name in certain cases.

But as-is is fine as well.

**Pull request merged**